swift/Lottie.swift

```diff
@@ -0,0 +1,169 @@
+    public convenience init(path: String) throws {
+        let animation = tvg_animation_new()
+        let picture = tvg_animation_get_picture(animation)
+
+        guard tvg_picture_load(picture, path) == TVG_RESULT_SUCCESS else {
+            throw LottieError.failedToLoadLottieFromPath
+        }
+
+        try self.init(animation: animation)
+    }
+
+    /// Initializes a new Lottie instance from a string representing the animation.
+    /// - Parameters:
+    ///   - string: The string representing the Lottie animation.
+    public convenience init(string: String) throws {
+        let animation = tvg_animation_new()
+        let picture = tvg_animation_get_picture(animation)
+
+        guard let cString = string.cString(using: .utf8),
+              tvg_picture_load_data(picture, cString, UInt32(cString.count), "lottie", "", false) == TVG_RESULT_SUCCESS
+        else {
+            throw LottieError.failedToLoadLottieFromString
+        }
+
+        try self.init(animation: animation)
+    }
+
+    /// Initializes the Lottie instance with frame count and duration extracted from the animation pointer.
+    private init(animation: Animation?) throws {
+        var numberOfFrames: Float = 0
+        tvg_animation_get_total_frame(animation, &numberOfFrames)
+
+        var duration: Float = 0
+        tvg_animation_get_duration(animation, &duration)
+
+        var width: Float = 0
+        var height: Float = 0
+        let picture = tvg_animation_get_picture(animation)
+        tvg_picture_get_size(picture, &width, &height)
+
+        self.animation = animation
+        self.numberOfFrames = Int(numberOfFrames)
+        self.duration = CMTime(seconds: Double(duration), preferredTimescale: 600)
+        self.size = CGSize(width: Double(width), height: Double(height))
+    }
+
+    /// Renders a Lottie frame into a buffer.
+    /// - Parameters:
+    ///   - index: The index of the frame to render.
+    ///   - buffer: The buffer to render the frame into.
+    ///   - stride: The stride of the buffer.
+    ///   - size: The desired size of the rendered frame.
+    ///   - crop: Optional rectangle to crop the rendered frame.
+    public func renderFrame(at index: Int, into buffer: Buffer, stride: Int, size: CGSize, crop: CGRect? = nil) throws {
+        guard index < numberOfFrames, index >= 0 else {
+            throw LottieError.frameIndexOutOfBounds
+        }
+
+        if canvas == nil {
+            canvas = try createCanvas(with: buffer, stride: stride, size: size)
+        }
+
+        prepareCanvasForRendering(frameIndex: index, size: size, crop: crop)
+        try renderCanvas()
+    }
+
+    /// Creates a new canvas for rendering, initializing the TVG engine and setting the target buffer.
+    private func createCanvas(with buffer: Buffer, stride: Int, size: CGSize) throws -> Canvas? {
+        guard tvg_engine_init(TVG_ENGINE_SW, UInt32(ProcessInfo.processInfo.activeProcessorCount)) == TVG_RESULT_SUCCESS else {
```

Comment:
I am not sure if `ProcessInfo.processInfo.activeProcessorCount` is what we want.

I found in the example, it has something like the following
```c++
//Threads Count
auto threads = std::thread::hardware_concurrency();
if (threads > 0) --threads;    //Allow the designated main thread capacity

if (Initializer::init(threads, CanvasEngine::Sw) != Result::Success) return false;
```

but we might need to fund out is this what really required? I guess for our task, `1` is fine?

I would spend more time digging into it, or even ask the clarification from the author + the Swift SPM support 😄

Reply:
Thanks for flagging this as a concern. 

Firstly, to note `ProcessInfo.processInfo.activeProcessorCount` gives us a similar output to `std::thread::hardware_concurrency()`, as it returns the number of active processors that we can run a thread on. 

But I think the larger concern you are getting at here is, do we need to use all of the active processors to render a single Lottie frame? And I think the even larger concern here is if that puts too much strain on the CPU. 

Going forward, I'm keen to trial out using the max processing power available, and will profile the results. Seeming the ThorVG documentation suggests to default to the available number of threads, I'm not too concerned, but it is valid that we flag this as potentially dangerous.